asv_bench/benchmarks/groupby.py

```diff
@@ -1,85 +1,108 @@
+        self.data.groupby(self.labels).apply(self.f)
 
-    def time_groupby_agg_builtins2(self):
-        self.df.groupby(['jim', 'joe']).agg([sum, min, max])
 
-#----------------------------------------------------------------------
-# dict return values
+class Apply(object):
 
-class groupby_apply_dict_return(object):
     goal_time = 0.2
 
     def setup(self):
-        self.labels = np.arange(1000).repeat(10)
-        self.data = Series(randn(len(self.labels)))
-        self.f = (lambda x: {'first': x.values[0], 'last': x.values[(-1)], })
+        N = 10**4
+        labels = np.random.randint(0, 2000, size=N)
+        labels2 = np.random.randint(0, 3, size=N)
+        self.df = DataFrame({'key': labels,
+                             'key2': labels2,
+                             'value1': np.random.randn(N),
+                             'value2': ['foo', 'bar', 'baz', 'qux'] * (N // 4),
+                             })
+        self.scalar_function = lambda x: 1
+
+    def time_scalar_function_multi_col(self):
+        self.df.groupby(['key', 'key2']).apply(self.scalar_function)
 
-    def time_groupby_apply_dict_return(self):
-        self.data.groupby(self.labels).apply(self.f)
+    def time_scalar_function_single_col(self):
+        self.df.groupby('key').apply(self.scalar_function)
 
+    @staticmethod
+    def df_copy_function(g):
+        # ensure that the group name is available (see GH #15062)
+        g.name
+        return g.copy()
+
+    def time_copy_function_multi_col(self):
+        self.df.groupby(['key', 'key2']).apply(self.df_copy_function)
+
+    def time_copy_overhead_single_col(self):
+        self.df.groupby('key').apply(self.df_copy_function)
 
-#----------------------------------------------------------------------
-# groups
 
 class Groups(object):
-    goal_time = 0.1
 
-    size = 2 ** 22
-    data = {
-        'int64_small': Series(np.random.randint(0, 100, size=size)),
-        'int64_large' : Series(np.random.randint(0, 10000, size=size)),
-        'object_small': Series(tm.makeStringIndex(100).take(np.random.randint(0, 100, size=size))),
-        'object_large': Series(tm.makeStringIndex(10000).take(np.random.randint(0, 10000, size=size)))
-    }
+    goal_time = 0.2
 
-    param_names = ['df']
+    param_names = ['key']
     params = ['int64_small', 'int64_large', 'object_small', 'object_large']
 
-    def setup(self, df):
-        self.df = self.data[df]
+    def setup_cache(self):
+        size = 10**6
+        data = {'int64_small': Series(np.random.randint(0, 100, size=size)),
+                'int64_large': Series(np.random.randint(0, 10000, size=size)),
+                'object_small': Series(
+                    tm.makeStringIndex(100).take(
+                        np.random.randint(0, 100, size=size))),
+                'object_large': Series(
+                    tm.makeStringIndex(10000).take(
+                        np.random.randint(0, 10000, size=size)))}
+        return data
```

Comment:
This still runs it for each benchmark in the class, instead of previously only once (I think)

Reply:
I think `setup_cache` will still only run once:

>If the setup is especially expensive, the setup_cache method may be used instead, which only performs the setup calculation once and then caches the result to disk. It is run only once also for repeated benchmarks and profiling, unlike setup

Reply:
yeah, the wording is not very clear. I interpreted it to not run between repetitions of the same benchmark, not for multiple benchmarks in the same class